plugins/base/src/test/kotlin/content/AllTypesPageTest.kt

```diff
@@ -0,0 +1,241 @@
+                                }
+                            }
+                        }
+                        group {
+                            link { +"sample.Test" }
+                            group {
+                                group {
+                                    +"Hello World!"
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    @Test
+    fun `all types isn't generated when not enabled by property`() {
+        testInline(
+            """
+            |/src/Test.kt
+            |package sample
+            |/**
+            | * Hello World!
+            | * 
+            | * Some other comment which should not be on All Types page
+            | */
+            |public class Test
+            |
+            |/**
+            | * Hello type
+            | */
+            |public typealias Alias = Int
+            """.trimIndent(),
+            configuration
+        ) {
+            pagesTransformationStage = { rootPage ->
+                assertNull(rootPage.allTypesPageNode())
+            }
+        }
+    }
+
+    @Test
+    fun `all types page isn't generated when there are NO types`() = withAllTypesPage {
+        testInline(
+            """
+            |/src/Test.kt
+            |package sample
+            |/**
+            | * Hello World!
+            | */
+            |public fun test() {}
+            """.trimIndent(),
+            configuration
+        ) {
+            pagesTransformationStage = { rootPage ->
+                assertNull(rootPage.allTypesPageNode())
+            }
+        }
+    }
+
+    @Test
+    fun `all types sorting depends only on simple name`() = withAllTypesPage {
+        testInline(
+            """
+            |/src/A.kt
+            |package bbb
+            |public class A
+            |/src/B.kt
+            |package xxx
+            |public class B
+            |/src/C.kt
+            |package aaa
+            |public class C
+            """.trimIndent(),
+            configuration
+        ) {
+            pagesTransformationStage = { rootPage ->
+                assertNotNull(rootPage.allTypesPageNode()).content.assertNode {
+                    group {
+                        header { +"root" } // module name
+                    }
+                    header { +"All Types" }
+                    table {
+                        group {
+                            link { +"bbb.A" }
+                        }
+                        group {
+                            link { +"xxx.B" }
+                        }
+                        group {
+                            link { +"aaa.C" }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    @Test
+    fun `all types description should be taken from most relevant sourceSet`() = withAllTypesPage {
```

Comment:
nitpick: looks like this test checks multiple cases at once, and it if it fails - it might be difficult to understand which case is broken, what this "most relevant" phrase means and what kind of fix to introduce. 

You can break it down into multiple tests that check for specific things. Basically, what you outlined in the comment:
```
//  1. if all data per source sets are the same - take it
//  2. if not, try to take common data
//  3. if not, try to take JVM data (as this is most likely to be the best variant)
//  4. if not, just take any data
```

Reply:
I've decided to postpone changing this test until final requirements will be discussed in https://github.com/Kotlin/dokka/pull/3267#discussion_r1390883283